Modify volume mounts SELinux labels on the fly based on :Z or :z

jessfraz opened the pull request:
carry of #5910

closes #5910

closes #13128

rhatdan commented:
@jfrazelle I updated my patch with some of the fixes required and also add the currently merged libcontainer patch so I could clean up my patch.

Please pull that into  your patch.

jfrazelle commented:
ah will update thanks @rhatdan

jfrazelle commented:
I will rebase this today, screwed myself over by merging the other volumes PR, facepalm

rhatdan commented:
I already re based, you might want to grab it.

LK4D4 commented:
@jfrazelle need rebase

jfrazelle commented:
yeah so I was going to rebase on top of michaels patch so save myself
headache

On Tue, May 5, 2015 at 3:28 PM, Alexander Morozov <notifications@github.com>
wrote:

> @jfrazelle <https://github.com/jfrazelle> need rebase
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/12572#issuecomment-99250505>.
>

rhatdan committed:
Modify volume mounts SELinux labels on the fly based on :Z or :z

This patch is extending the qualifiers on the -v command to allow
an admin to tell the system to relabel, content. There might be a
need for something similar for changing the DAC Permissions.

Signed-off-by: Jessica Frazelle <princess@docker.com>

rhatdan commented:
I have updated my original pull requests.  Any word on this being merged.

jfrazelle commented:
we are trying to get it in :) I will take a look

On Tue, May 12, 2015 at 4:52 AM, Daniel J Walsh <notifications@github.com>
wrote:

> I have updated my original pull requests. Any word on this being merged.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/12572#issuecomment-101249865>.
>

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "selinux-labels-carry" git@github.com:jfrazelle/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

This will update the existing PR, so **DO NOT** open a new one.

jfrazelle commented:
ok i updated the pr here with a few changes

icecrime commented:
@rhatdan Sorry for the late question, but any reason not to *always* relabel when SELinux is enabled? Said otherwise: do we really need the `:zZ` syntax? (cc @crosbymichael)

rhatdan commented:
How would docker know whether or not to relabel?  If I want a shared image or an private image, If SELinux is disabled in docker, if I am running a privileged container.  If I volume mount in content that the container can read, no reason to change the label.  For example -v /usr/bin:/usr/bin

rhatdan commented:
In certain cases we definitely do not want to relabel. For example mounting /home/dwalsh into a container could cause lots of problems.

icecrime commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -2178,6 +2178,19 @@ mount the volumes in read-only or read-write mode, respectively. By default,
 the reference container.
 
+Labeling systems like SELinux require proper labels be placed on volume content
+mounted into a container, otherwise the secuirty system might prevent the
```
security*

icecrime commented:
Alright, makes sense.

LGTM

jfrazelle commented:
typo fixed

pdevine commented:
I realize this is probably out of scope for the release, but ideally wouldn't we save the existing selinux context and restore it when the bind mount/container is torn down?  The problem right now is that if you use the :zZ option the relabel happens, the context changes and then the next time you go to mount that volume into the container it doesn't need the :zZ option and you may have ended up in the situation @rhatdan is describing.  Mounting your home dir into a container shouldn't screw up the context permanently.

rhatdan commented:
If you want to restore the context to a safe context, it would probably be better to do a restorecon on the content rather then saving the labels, since there could be several labels that get changed.

restorecon -R -v PATH sets the labels back to the system default.

I could build a patch to do this but please do not hold up this patch set waiting for this.  This patch set in one form or another is over a year old.

rhatdan commented:
One problem with this also is if you label the content with a z, it is for a shared container.  It would be difficult to figure out whether all containers are no longer using the volume.

pdevine commented:
I'm fairly certain we should be able to detect if the file/directory was still bind mounted and then not try to restore the selinux context if that was the case.

LK4D4 commented on daemon/volumes.go:
```diff
@@ -164,11 +177,25 @@ func (container *Container) registerVolumes() {
+		defer resetLabeling()
+
+		if err := label.SetFileCreateLabel(container.MountLabel); err != nil {
+			logrus.Debugf("Unable to setup default labeling for volume creation %s: %v", path, err)
```
I think it is at least `Warn`, but probably `Error` or returning error, because it is probably security issue.

jfrazelle replied:
will update :)

jfrazelle commented:
updated

icecrime commented:
Still LGTM! Ping @LK4D4.

LK4D4 commented:
@jfrazelle Need rebase :)

LK4D4 commented on daemon/volumes.go:
```diff
@@ -164,11 +177,25 @@ func (container *Container) registerVolumes() {
 			continue
 		}
+		if err := resetLabeling(); err != nil {
+			logrus.Debugf("Unable to reset labeling %s: %v", path, err)
```
I'm personally think that `resetLabeling` should panic on error :) But at least `Errorf` will be okay for me. Also in `defer` too :)

jfrazelle commented:
Will do tomorrow :) I assume because volume refactor ;)

On Saturday, May 23, 2015, Alexander Morozov <notifications@github.com>
wrote:

> @jfrazelle <https://github.com/jfrazelle> Need rebase :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/12572#issuecomment-104977254>.
>

jessfraz committed:
apply selinux labels volume patch on volumes refactor

Signed-off-by: Jessica Frazelle <princess@docker.com>

jfrazelle commented:
basically had to rewrite a lot of it for the volumes refactor, so this needs a review, ping @rhatdan @LK4D4 @icecrime

cpuguy83 commented on daemon/volumes.go:
```diff
@@ -6,11 +6,13 @@ import (
 	"github.com/docker/docker/daemon/execdriver"
 	"github.com/docker/docker/pkg/chrootarchive"
 	"github.com/docker/docker/runconfig"
+	"github.com/docker/docker/vendor/src/github.com/docker/libcontainer/label"
```
Whoops

cpuguy83 commented on daemon/volumes.go:
```diff
@@ -205,6 +233,17 @@ func (daemon *Daemon) registerMountPoints(container *Container, hostConfig *runc
+			runtime.LockOSThread()
+			//  defer reset the label
+			defer label.SetFileCreateLabel("")
+			defer runtime.UnlockOSThread()
```
This defer is in a for-loop, and probably just needs to wrap the SetFileCreateLabel (below) anyway.

cpuguy83 commented on daemon/volumes.go:
```diff
@@ -205,6 +233,17 @@ func (daemon *Daemon) registerMountPoints(container *Container, hostConfig *runc
 		if len(bind.Name) > 0 && len(bind.Driver) > 0 {
+			runtime.LockOSThread()
+			//  defer reset the label
+			defer label.SetFileCreateLabel("")
```
This should not be in the for-loop.

icecrime commented:
Reviewed: agree with all @cpuguy83 remarks (nice catches!), none to add.

jfrazelle commented:
updated

cpuguy83 commented:
Janky is sad panda.

rhatdan commented:
I believe this is wrong.
 +		bind.RW = mode == "rw" 

Should probably be.
bind.RW = rwModes(mode)

jfrazelle commented:
@rhatdan good catch, updated

jfrazelle commented:
hang on updating to lock the entire range... https://github.com/jfrazelle/docker/commit/4c7371667eeeda380c3f7fbc6d3cdcba2ad09564#commitcomment-11373232

icecrime commented on daemon/volumes.go:
```diff
@@ -205,11 +236,22 @@ func (daemon *Daemon) registerMountPoints(container *Container, hostConfig *runc
+			// create the volume
 			v, err := createVolume(bind.Name, bind.Driver)
 			if err != nil {
 				return err
```
How bad would it be to return on this error case without the label being reset? (cc @rhatdan)

jfrazelle replied:
bad let me fix

On Tue, May 26, 2015 at 2:41 PM, Arnaud Porterie <notifications@github.com>
wrote:

> In daemon/volumes.go
> <https://github.com/docker/docker/pull/12572#discussion_r31082773>:
>
> > @@ -205,11 +236,22 @@ func (daemon *Daemon) registerMountPoints(container *Container, hostConfig *runc
> >  		}
> >
> >  		if len(bind.Name) > 0 && len(bind.Driver) > 0 {
> > +			// set the label
> > +			if err := label.SetFileCreateLabel(container.MountLabel); err != nil {
> > +				return fmt.Errorf("Unable to setup default labeling for volume creation %s: %v", bind.Source, err)
> > +			}
> > +
> > +			// create the volume
> >  			v, err := createVolume(bind.Name, bind.Driver)
> >  			if err != nil {
> >  				return err
>
> How bad would it be to return on this error case without the label being
> reset? (cc @rhatdan <https://github.com/rhatdan>)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/12572/files#r31082773>.
>

rhatdan replied:
On 05/26/2015 05:41 PM, Arnaud Porterie wrote:
>
> In daemon/volumes.go
> <https://github.com/docker/docker/pull/12572#discussion_r31082773>:
>
> > @@ -205,11 +236,22 @@ func (daemon *Daemon) registerMountPoints(container *Container, hostConfig *runc
> >  		}
> >  
> >  		if len(bind.Name) > 0 && len(bind.Driver) > 0 {
> > +			// set the label
> > +			if err := label.SetFileCreateLabel(container.MountLabel); err != nil {
> > +				return fmt.Errorf("Unable to setup default labeling for volume creation %s: %v", bind.Source, err)
> > +			}
> > +
> > +			// create the volume
> >  			v, err := createVolume(bind.Name, bind.Driver)
> >  			if err != nil {
> >  				return err
>
> How bad would it be to return on this error case without the label
> being reset? (cc @rhatdan <https://github.com/rhatdan>)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/12572/files#r31082773>.
>
It could cause problems. Basically the call to SetFileCreateLabel(LABEL)

Tells the kernel that all file objects created by this process thread
should be labeled LABEL.  Until it is reset to "", which tells the
kernel to go back to the default.  We should defer setting the greate
label to "" which is what my code originally did.

icecrime commented:
Code LGTM. Ping @cpuguy83 for another review!

cpuguy83 commented on daemon/volumes.go:
```diff
@@ -107,23 +117,42 @@ func parseVolumesFrom(spec string) (string, string, error) {
+	return roModes[mode] || rwModes[mode]
 }
 
 func (container *Container) specialMounts() []execdriver.Mount {
```
These aren't being used anywhere and looks like it should have been removed (or just renamed) in the volumes refactor as `networkMounts` has now replaced it.

cpuguy83 replied:
And just to be clear I think these label calls need to into `networkMounts` and we can remove `specialMounts` since it shouldn't exist anyway.

cpuguy83 commented on daemon/volumes.go:
```diff
@@ -205,11 +237,26 @@ func (daemon *Daemon) registerMountPoints(container *Container, hostConfig *runc
 			if err != nil {
+				// reset the label
+				if e := label.SetFileCreateLabel(""); e != nil {
+					logrus.Errorf("Unable to reset labeling for volume creation %s: %v", bind.Source, e)
```
Instead of logging should this just be handled by the caller? We can wrap the returned error with this added context.

jfrazelle commented:
updated.... and i like the log

cpuguy83 commented:
Ok, testing on CentOS I see there's a problem (and a regression from volume refactor).

In this PR, labels are only set for when a driver is specified for the bind, so... `docker run -v <name>:/path` instead of for all binds (ie, `docker run -v /test:/test`).

Also regression is that the vfs graphdriver was automatically setting labels so `s0`, the `local` driver does not and so volumes created by the `local` driver are not accessible at all.

calavera commented on daemon/volumes_experimental_unit_test.go:
```diff
@@ -34,28 +34,29 @@ func TestGetVolumeDriver(t *testing.T) {
+		{"name:/tmp", "", "/tmp", "", "name", "local", "", true, false},
+		{"name:/tmp", "external", "/tmp", "", "name", "external", "", true, false},
+		{"name:/tmp:ro", "local", "/tmp", "", "name", "local", "", false, false},
+		{"local/name:/tmp:rw", "", "/tmp", "", "local/name", "local", "", true, false},
```
I think it would be nice to have new RW modes in this table.

tiborvass commented:
Code LGTM

rhatdan commented:
I think @cpuguy83 is working through a problem

tiborvass commented:
@rhatdan huh, thanks for the reminder!

cpuguy83 committed:
Use SELinux labels for volumes

Fixes a regression from the volumes refactor where the vfs graphdriver
was setting labels for volumes to `s0` so that they can both be written
to by the container and shared with other containers.
When moving away from vfs this was never re-introduced.
Since this needs to happen regardless of volume driver, this is
implemented outside of the driver.

Fixes issue where `z` and `Z` labels are not set for bind-mounts.

Don't lock while creating volumes

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "selinux-labels-carry" git@github.com:jfrazelle/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~4
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

This will update the existing PR, so **DO NOT** open a new one.

calavera commented on daemon/daemon.go:
```diff
@@ -1221,16 +1221,16 @@ func (daemon *Daemon) verifyHostConfig(hostConfig *runconfig.HostConfig) ([]stri
 		return err
 	}
 
+	if err := daemon.registerMountPoints(container, hostConfig); err != nil {
```
this was outside the lock in purpose actually. We didn't want to block the container when volumes are created to avoid blocking docker ps due to external connections. Is there a reason to move it back here?

/cc @LK4D4

cpuguy83 replied:
Because parseSecurityOpt needs to happen first.
But also registerMountPoints is manipulating the container struct.

calavera replied:
sounds good :+1:

cpuguy83 replied:
Which part?

calavera replied:
both parts :smile:

cpuguy83 replied:
@jfrazelle https://github.com/cpuguy83/docker/commit/e83230bbd34946dc7189b363d4ce95c786b14dc1 to fix locking

calavera replied:
:heart_eyes_cat:

jfrazelle replied:
cool will apply and force push

calavera commented:
@cpuguy83 it looks like there are a bunch of errors around volumes in experimental. I'm not completely sure why.

jfrazelle commented:
updated w feedback from @rhatdan @cpuguy83 @calavera PTAL

calavera commented:
code LGTM, thanks @jfrazelle and @cpuguy83 !

jfrazelle commented:
omg im having deja vu

icecrime commented:
Pushing this to docs review, please @rhatdan give it a final look if you have a chance!

icecrime commented:
Ping @moxiegirl @SvenDowideit!

jfrazelle commented:
this already had docs review just fyi

jfrazelle commented:
on the old one

icecrime commented:
Right, sorry @jfrazelle. Let's merge this :tada:

moxiegirl commented:
Anyone have an example of what the "shared content label" looks like --- is it visible through a command?

moxiegirl commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -2181,6 +2181,19 @@ mount the volumes in read-only or read-write mode, respectively. By default,
+Labeling systems like SELinux require proper labels be placed on volume content
+mounted into a container, otherwise the security system might prevent the
+processes running inside the container from using the content. By default,
+volumes are not relabeled.
```
How did the concept of "relabeling" enter into this?  Or do you mean you can only label a volume once and after that you can't relabel it?

cpuguy83 replied:
This means that docker does not by default changing the labels set by the OS.

moxiegirl replied:
Thanks @cpuguy83 cleared that right up.

moxiegirl commented on docs/man/docker-run.1.md:
```diff
@@ -396,6 +396,21 @@ used in other containers using the **--volumes-from** option.
+volumes are not relabeled.
+
+Two suffixes :z or :Z can be added to the volume mount. These suffixes tell
+Docker to relabel file objects on the shared volumes. The 'z' option tells
```
relabel > label?

cpuguy83 replied:
relabel is correct. These are labels that selinux sets, and docker is changing that label.

rhatdan replied:
Shared label looks something like
system_u:object_r:svirt_sandbox_file_t:s0

Private label looks like
system_u:object_r:svirt_sandbox_file_t:s0:c1,c2

Where c1,c2 is different for every container. Has a range of c1-c1024.

rhatdan commented:
On 05/27/2015 06:06 PM, moxiegirl wrote:
>
> Anyone have an example of what the "shared content label" looks like
> --- is it visible through a command?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/12572#issuecomment-106092090>.
>
Files get labeled directly by the kernel, if you change the label, we
refer to that to relabeling, or changing the label.

moxiegirl commented:
Thanks @rhatdan one follow on question this statement "By default, volumes are not relabeled." Is it correct then that I have this going on conceptually

Volume + label from host
   |
------------container--------
   |
   |____content + label

rhatdan commented:
Yes by default no label changes happen.  Only if the user specified z or  Z.  If they specify a container volume that gets created in vfs directory it will get created with the shared label, since it is assumed that container volumes will be shared between containers.

derekstavis commented:
So whats the syntax for multiple options?
`-v /mnt/uploads:/mnt/uploads:rw,z`?
`-v /mnt/uploads:/mnt/uploads:rw:z`?

cpuguy83 commented:
@derekstavis I believe it's the comma syntax. valid modes: https://github.com/docker/docker/blob/master/volume/volume.go#L33-L50

derekstavis commented:
Oh, thanks @cpuguy83. I'm pretty new to Docker, this options aren't documented, are them in experimental?

derekstavis commented:
Looks like it's not working. Here comes some debug info:

Host machine mount point:
```
$ mount
...
/dev/sdb1 on /run/storage type ext4 (rw,noexec,nosuid,nodev,mode=0755)
...
```

Docker volume flag:
```
-v /run/storage:/mnt/uploads:rw,z
```

Container mount point:
```
$ sudo docker exec 6ed0e172b7ac mount
...
/dev/sdb1 on /mnt/uploads type ext4 (ro,relatime,errors=remount-ro,data=ordered)
...
```

cpuguy83 commented:
@derekstavis You are correct, it doesn't appear to be properly documented.
I opened #15896 to track

`z` is not a real mount flag, only a docker option.
The way to test would be to do `ls -z /run/storage`. It would have a label attached to it which allows a containerized process to write to it.

derekstavis commented:
This is the output of `ls -Z` on both machines:

On the host machine:
```
$ ls -Z /run/storage
? lost+found
? test
```

On the container:
```
$ sudo docker exec 6ed0e172b7ac ls -Z /mnt/uploads
? lost+found
? test
```

cpuguy83 commented:
@derekstavis This would need to be done on the host size, not in the container.
Can we move this to IRC? Happy to help there. Keep in mind every maintainer gets an email with every comment on GH.

derekstavis commented:
Yes, we can. I'm __dereks__. Sorry for the buzz.

chadfurman commented:
Is this CLI only?

cpuguy83 commented:
@chadfurman It's in the API. I would suggest that you modify these labels manually, though.
